### DISCUSSION TITLE
Too many emojis in 'Concurrency and async / await' explanation

### DISCUSSION THREAD

**Discussion by kittenswolf** [fastapi/fastapi#6656] [Questions] [answered]
The explanation page for concurrency, and more specifically its example with burgers (https://fastapi.tiangolo.com/async/#concurrent-burgers) becomes hard to read for me because of its many emojis.

An excerpt:

> You go with your crush 😍 to get fast food 🍔, you stand in line while the cashier 💁 takes the orders from the people in front of you.
> Then it's your turn, you place your order of 2 very fancy burgers 🍔 for your crush 😍 and you.
> You pay 💸.
> The cashier 💁 says something to the guy in the kitchen 👨‍🍳 so he knows he has to prepare your burgers 🍔 (even though he is currently preparing the ones for the previous clients).
> The cashier 💁 gives you the number of your turn.
> While you are waiting, you go with your crush 😍 and pick a table, you sit and talk with your crush 😍 for a long time (as your burgers are very fancy and take some time to prepare ✨🍔✨).
> As you are sitting on the table with your crush 😍, while you wait for the burgers 🍔, you can spend that time admiring how awesome, cute and smart your crush is ✨😍✨.
...
But as you go away from the counter and sit on the table with a number for your turn, you can switch 🔀 your attention to your crush 😍, and "work" ⏯ 🤓 on that. Then you are again doing something very "productive" 🤓, as is flirting with your crush 😍.

In my opinion, the many emojis don't add anything to the explanation and make it hard to read. **Every emoji breaks the flow when reading because my eyes are distracted by the emojis' size and colour.**

Is the '😍' after every 'crush' necessary? Same with burger and 🍔, cashier and 💁 and so on.

I propose cutting down on or completely removing emojis from the documentation to help with readability.

The same issue, though less pronounced, is also apparent in the section 'Asynchronous Code' (https://fastapi.tiangolo.com/async/#asynchronous-code).

***

(If you want to know what it feels like to read the documentation, try reading this:)

In my 🧑opinion 🤔, the many🔢 emojis don't add anything😣 to the explanation😪 and make it hard to read😓. Every emoji🤪 breaks the flow☄🌬 when reading📖 because my eyes👀👀 are distracted😜😒 by the emojis' size🥵 and colour🤡.

**Comment by Kludex** [comment]
Emojis are part of our culture. :sweat_smile:

> (If you want to know what it feels like to read the documentation, try reading this:)

I've read the documentation, and I don't feel the same as reading your example, as it's clearly ill-intentioned to prove your point. :kissing: 

That being said, we can wait to see if people have the same opinion as you, and/or you can propose a PR changing that page, and @tiangolo you'll check it at some point.

I don't think this change is needed, nor do I feel harmed by those emojis, sometimes I even get the opposite effect as the one you mentioned. Which I think is because humans tend to relate images better than letters. 

EDIT: Well... Months have passed, and my opinion has changed. I no longer have strong feelings about this. Although, I think the most important argument here is the one mentioned [here](https://github.com/tiangolo/fastapi/issues/3273#issuecomment-1018297338). :wave:

**Comment by baby5** [comment]
you can write non emojis one in your self page.

**Comment by ParitoshSingh07** [comment]
That page was painful to read. Can the emojis be toned down please? 

For a middle ground, perhaps introduce them in the opening sentence, and then stick to using them at the end of sentences, avoiding the temptation to use them mid-sentences outside their first occurrence.

**Comment by kms70847** [comment]
"we can wait to see if people have the same opinion as you". It's been a few months, which seems like a reasonable time frame for waiting and seeing. The original post has 14 thumbs up and 2 thumbs down, so it seems like the community is in favor of it.

**Comment by sanchesper** [comment]
Found this post because I made a search on Google: "why fastapi doc full of emojis". I didn't understand why the words are not replaced by the emojis, but instead have the word and the emoji together. If want to keep the emojis, so remove the redundancy for better reading. (i.e. "You go with your 😍 to get 🍔, you stand in line while the 💁 takes the orders from the people in front of you.")

**Comment by v3ss0n** [comment]
May be targeted audience is 16 year olds.
J/K.

**Comment by Zeta611** [comment]
When I landed on the FastAPI documentation page explaining concurrency last week, I almost immediately returned to the previous page.
I even stumbled upon this issue page after googling `concurrency emoji`--I was looking for an emoji that fits with the concept *concurrency* for my blog post.

The reason I left the doc was mainly because I already knew the concept, but I would have at least *tried* to read it to see how it explains concurrency if it were not for the emojis.
And even if I were new to the concept, I probably would have searched for an alternative source explaining it.

IMHO, the use of emojis 🔀 and ⏯ was actually pretty smart, and perhaps the use of 😍s has lightened the mood... But the current 3😀/sentence is definitely not helping the readability.

**Comment by joe-butler2** [comment]
cope

**Comment by acidjunk** [comment]
or not to cope

**Comment by chocabloc** [comment]
that is the question

**Comment by YodaEmbedding** [comment]
The 🎡  of  🕦  turns, and ages come and pass, leaving emojis that become legend.  ✨✨✨  Legends  ✨✨✨  f a d e  to myth,  🐉  and even myth 🐲  is l o n g forgotten when the Emoji that gave it birth 🧑‍🍼👩‍🍼  comes again. ⏳ In one such Emoji, called the 🏎️ 🏎️ 🏎️ Fast API 🏎️ 🏎️ 🏎️  by some, an Emoji yet to come, an Emoji long past, a wind 🪁 🪁  rose in the  ⛰️ ⛰️ ⛰️  Misty Mountains ⛰️ ⛰️ ⛰️  of Emijiji. The wind 🪁  was not the beginning ⏳.  👉 There are neither beginnings nor endings to the turning of the 🎡  of 🕡.  But it was a beginning. ⏳

**Comment by intrnl** [comment]
Another major problem with this which isn't being talked enough is that for people who have reading/visual impairments and thus needs to rely on screen readers, the overuse of emojis adds a lot of noise that otherwise distracts from the actual material.

https://user-images.githubusercontent.com/20620901/150496694-7a6f333e-2ce0-409a-a656-f7d880381354.mp4

Accessibility is important.

**Comment by vimpostor** [comment]
I like the irony of reacting to the original issue above with all sorts of different emoji. :eyes:

**Comment by tsirolnik** [comment]
Emojis are 🐙😍 a 😯🍇 beautiful represnetial addition ➕ to 🙅 our 👨📸 languages, they provide an ability to 💦 share more ➕💦 deep 😩 context 😏😎 than 👄♀ just written 📝📝 words, ☺ I see no problems with 🅱😶 emojis in 👈 written 📝 text, 📲 as 👦 said before, they 😱 provide 💰😡 deeper 😩😩 context and vibrant colors to 💨👬 the 🔚 mundane repetitive language. ℹ Emojis 👵😤 allow 👍 an 👹🐎 inclusive conversation, 🤓🤓 something 🐴😮 that 😔 wasn't available 💢✔ before. Emojis 🅱👵 can 💦🙅 pack more 🍗 meaning in 💁😅 less text. 📲 If ☝ emojis hurt 🤕🤕 you then you 🕎💃 might ♂⚰ need to 😢 check your privilege. It's 22 people grow up.

**Comment by snctfd** [comment]
I agree with the OP, this is fairly atrocious to read, not to mention the issue of accessibility brought up by @intrnl.

**Comment by oz123** [comment]
Emojis are hard to read on a Braille keyboard.
And frankly, the ABC was invented to replace hieroglyphics for simplicity.
Sprinkling emojis here and there is fine, but boy, too much is just too much.

~~The suggestion by @jwilber bellow should be a firefox\chrome plugin!!! Anyone up to the mission?~~

Here is a FF\Chrome plugin:
https://github.com/erikdesjardins/no-emoji

**Comment by jwilber** [comment]
If emojis are part of your culture, maybe give them all a class controlling visibility/rendering that can easily be toggled on/off by some widget at the top of each article? 

Basically light/dark mode, but for emojis. 

👁 👄 👁

**Comment by nicktimko** [comment]
Back in my day, we didn't have these fancy-pants characters. We made do with 7-bit code points and were happy about it! <3

**Comment by logan-pairity** [comment]
Please think about the screen readers.

**Comment by ghost** [comment]
While not a real solution, I for one found this made the documentation easier to read.

Using Chrome 97, I opened the developer tools, and entered this into the console while in the page and ran it (untested on other browsers):

`document.body.innerHTML = document.body.innerHTML.replace(/(?:[\u2700-\u27bf]|(?:\ud83c[\udde6-\uddff]){2}|[\ud800-\udbff][\udc00-\udfff]|[\u0023-\u0039]\ufe0f?\u20e3|\u3299|\u3297|\u303d|\u3030|\u24c2|\ud83c[\udd70-\udd71]|\ud83c[\udd7e-\udd7f]|\ud83c\udd8e|\ud83c[\udd91-\udd9a]|\ud83c[\udde6-\uddff]|\ud83c[\ude01-\ude02]|\ud83c\ude1a|\ud83c\ude2f|\ud83c[\ude32-\ude3a]|\ud83c[\ude50-\ude51]|\u203c|\u2049|[\u25aa-\u25ab]|\u25b6|\u25c0|[\u25fb-\u25fe]|\u00a9|\u00ae|\u2122|\u2139|\ud83c\udc04|[\u2600-\u26FF]|\u2b05|\u2b06|\u2b07|\u2b1b|\u2b1c|\u2b50|\u2b55|\u231a|\u231b|\u2328|\u23cf|[\u23e9-\u23f3]|[\u23f8-\u23fa]|\ud83c\udccf|\u2934|\u2935|[\u2190-\u21ff])/g, '')`

[JS Source from stack exchange](https://apple.stackexchange.com/questions/182967/how-to-disable-emoji-i-see-sometimes-while-browsing#answer-316956)

## Before the script

![Before the script](https://user-images.githubusercontent.com/79225093/151870348-051418bc-8900-4860-bc54-bcf7b2cd715c.png)


## After the script

![After the script](https://user-images.githubusercontent.com/79225093/151870346-003b11ee-3867-4ee8-be44-fab6017f727d.png)

**Comment by mattyoungberg** [comment]
The original issue post has 300+ positive reactions. It's time to address this. 

Emojis at the end of sentences, IMO, can be used _sparingly_, but CERTAINLY not in the middle of them. You're wrecking your ability to communicate crucial information about how FastAPI works over something as trivial and petty as emoji use. 

Make docs readable again. Process the pull request here from the user who opened this issue in the first place (#3893).

**Comment by v3ss0n** [comment]
Well the guy already abandoned the project.

**Comment by kittenswolf** [comment]
@tiangolo Can you give some input on this issue? Cheers

**Comment by UP929312** [comment]
Does @tiangolo even work on this anymore? Why call it open source when he gatekeeps the code...

**Comment by chbndrhnns** [comment]
@UP929312 i think yes: https://youtu.be/8IJkSs9Dvjo?t=48m30s

**Comment by abcdenis** [comment]
It is possible to write a script that will replace emoji with words before reading :)
https://stackoverflow.com/a/57580714/116373

**Comment by girisagar46** [comment]
https://fastapi.tiangolo.com/async/#concurrent-burgers

emojis are now replaced with illustrative images

**Comment by tiangolo** [comment, marked as answer]
It's sad that it seems some didn't realize the purpose of the emojis there.

The emojis in that section were not intended as plainly decorative. I agree it feels too much and I didn't like how dense it felt. But they had a purpose.

The emojis in the article were intended as a [visual cue](https://en.wikipedia.org/wiki/Sensory_cue), to make it easier to relate a reference to one entity with other references to the same entity. They were always the same emoji, the same number (when repeated). It was all to help reduce the [cognitive load](https://en.wikipedia.org/wiki/Cognitive_load) while studying the concepts.

That way, when you see an emoji, you can do the mental equivalent of a hash key search (a Python dict lookup), in $O(1)$ (constant time), matching the emoji with the entity. Instead of having to parse again the whole text searching for all the references to try and see if that particular word refers to the burger, the crush, the cook, etc. which would be $O(n)$, linear time (like searching for a value in a list).

Also, thumbs up in this issue only count for the people that don't like emojis and come looking for the issue. But for the people that are fine with emojis, they wouldn't come looking for the issue to add a thumbs down, so it's not really a safe/fair measure.

On the other hand, that section in the docs is what I have seen screenshotted the most, with people publishing how much they like it. So, I guess it's just a division in taste.

---

I don't appreciate sarcastic, passive-aggressive, and demeaning comments, those are not welcome here.

Thanks to everyone here that made comments to support, improve, and help, with clear and calm arguments. That is much appreciated.

---

I recently hired Ketrina Yim, a professional illustrator specialized in computer science to add illustrations to that section. That was always my ideal, and the emojis were just trying to help while that was available, it has finally landed.

You can see the new docs here: https://fastapi.tiangolo.com/async/#concurrency-and-burgers

With that, I reduced the use of emojis a lot. But I still kept some in some places that don't have illustrations and where the visual cues to reduce cognitive load would make sense.

Given that, I'm gonna close this issue now. Thanks for the interactions everyone! :coffee:
